plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -823,22 +829,91 @@ compileExpr e = traceCompilation 2 ("Compiling expr:" GHC.<+> GHC.ppr e) $ do
+    -- See Note [Coercions and newtypes]
+    GHC.Cast body _ -> compileExpr body
+    GHC.Type _ -> throwPlain $ UnsupportedError "Types as standalone expressions"
+    GHC.Coercion _ -> throwPlain $ UnsupportedError "Coercions as expressions"
+
+compileCase ::
+  (CompilingDefault uni fun m ann) =>
+  -- | Whether the variable is dead in the expr
+  (GHC.Var -> GHC.CoreExpr -> Bool) ->
+  -- | Whether we should try to rewrite opaque constructor applications
+  Bool ->
+  BuiltinsInfo uni fun ->
+  GHC.CoreExpr ->
+  GHC.Var ->
+  GHC.Type ->
+  [GHC.CoreAlt] ->
+  m (PIRTerm uni fun)
+compileCase isDead rewriteOpaque binfo scrutinee binder t alts = case alts of
+  [GHC.Alt con bs body]
     -- See Note [Evaluation-only cases]
-    GHC.Case scrutinee b _ [GHC.Alt _ bs body] | all (GHC.isDeadOcc . GHC.occInfo . GHC.idInfo) bs -> do
+    | all (`isDead` body) bs -> do
       -- See Note [At patterns]
       scrutinee' <- compileExpr scrutinee
-      withVarScoped b $ \v -> do
+      withVarScoped binder $ \v -> do
         body' <- compileExpr body
         -- See Note [At patterns]
         let binds = [PIR.TermBind annMayInline PIR.Strict v scrutinee']
         pure $ PIR.mkLet annMayInline PIR.NonRec binds body'
-    GHC.Case scrutinee b t alts -> do
+    | rewriteOpaque
+    , GHC.DataAlt dataCon <- con
+    , let dataConName = GHC.dataConName dataCon
+    , let dataConModule = GHC.moduleNameString . GHC.moduleName <$> GHC.nameModule_maybe dataConName
+    , let dataConBase = GHC.occNameString (GHC.nameOccName dataConName)
+    , (dataConModule, dataConBase) `elem` opaqueTypes -> do
+        -- Attempt to rewrite opaque constructors, since opaque constructors cannot be compiled.
+        -- For example, this rewrites
+
+        -- ```
+        -- case scrut of b {BuiltinList xs} -> ...BuiltinList @BuiltinData xs...
+        -- ```
+        --
+        -- into
+        --
+        -- ```
+        -- case scrut of b {BuiltinList xs} -> ...b...
+        -- ```
+        --
+        -- after which `xs` is hopefully dead, and we can then compile it using the
+        -- `all (`isDead` body) bs` branch of `compileCase`.
+        let f (GHC.collectArgs -> (GHC.Var (GHC.isDataConId_maybe -> Just dataCon'), args0))
+              | dataCon == dataCon'
+              -- Discard type arguments
+              , let args = mapMaybe (\case GHC.Var v -> Just v; _ -> Nothing) args0
+              , length bs == length args
+              , all (uncurry (==)) (bs `zip` args) =
+                GHC.Var binder
+            f other = other
+            -- This time we can no longer use `GHC.isDeadOcc`. Instead we check manually.
+            isDead' b = not . any (== b) . universeBi
+        -- If some binders are still alive, we have to give up (rather than trying to rewrite
+        -- opaque constructor applications again, which will loop), hence `False`.
+        compileCase isDead' False binfo scrutinee binder t [GHC.Alt con bs (transform f body)]
```

Comment:
I don't love the recursion here, since it makes me think about whether or not we can re-enter the same branch. I'm sure it's fine, but I'd be tempted to structure it as a two-phase thing.

Reply:
Please elaborate how a two-phase approach looks like.

Reply:
I meant, have a version that clearly does do the rewriting and then one that clearly doesn't. But I think you achieve that with the boolean flag, I just hadn't seen that you pass it as False here.